diff -ur -N -d -x CMakeLists.txt -x tags wolfssl-3.14.0-pristine/IDE/MYSQL/CMakeLists_wolfSSL.txt wolfssl-3.14.0/IDE/MYSQL/CMakeLists_wolfSSL.txt
--- wolfssl-3.14.0-pristine/IDE/MYSQL/CMakeLists_wolfSSL.txt	2018-03-02 22:38:23.000000000 +0100
+++ wolfssl-3.14.0/IDE/MYSQL/CMakeLists_wolfSSL.txt	2018-05-22 16:24:14.711987854 +0200
@@ -32,7 +32,7 @@
 
 SET(WOLFSSL_SOURCES  src/crl.c src/internal.c src/keys.c src/sniffer.c
         src/tls.c src/wolfio.c src/ocsp.c src/ssl.c
-                                ../../client/get_password.c )
+                                ../../client/get_password.cc )
 
 ADD_CONVENIENCE_LIBRARY(wolfssl ${WOLFSSL_SOURCES})
 
diff -ur -N -d -x CMakeLists.txt -x tags wolfssl-3.14.0-pristine/src/internal.c wolfssl-3.14.0/src/internal.c
--- wolfssl-3.14.0-pristine/src/internal.c	2018-03-02 22:39:08.000000000 +0100
+++ wolfssl-3.14.0/src/internal.c	2018-05-22 16:24:14.716987936 +0200
@@ -3993,6 +3993,17 @@
     ssl->devId = ctx->devId;
 #endif
 
+    /* downgrade the version according to the flags */
+    if (ssl->options.downgrade)
+    {
+      if ((ssl->options.mask & (SSL_OP_NO_SSLv3 | SSL_OP_NO_TLSv1)) == SSL_OP_NO_TLSv1)
+        wolfSSL_SetVersion(ssl, WOLFSSL_SSLV3);
+      else if ((ssl->options.mask & (SSL_OP_NO_TLSv1 | SSL_OP_NO_TLSv1_1)) == SSL_OP_NO_TLSv1_1)
+        wolfSSL_SetVersion(ssl, WOLFSSL_TLSV1);
+      else if ((ssl->options.mask & (SSL_OP_NO_TLSv1_1 | SSL_OP_NO_TLSv1_2)) == SSL_OP_NO_TLSv1_2)
+        wolfSSL_SetVersion(ssl, WOLFSSL_TLSV1_1);
+    }
+
     if (writeDup == 0) {
         int keySz = 0;
 #ifndef NO_CERTS
@@ -17270,6 +17281,7 @@
         }
         else
 #endif
+#ifndef WOLFSSL_MYSQL_COMPATIBLE
         if (ssl->ctx->method->version.major == SSLv3_MAJOR &&
                              ssl->ctx->method->version.minor == TLSv1_2_MINOR) {
             /* TLS v1.2 capable client not allowed to downgrade when connecting
@@ -17282,6 +17294,7 @@
                 return VERSION_ERROR;
             }
         }
+#endif
 
         /* session id */
         ssl->arrays->sessionIDSz = input[i++];
diff -ur -N -d -x CMakeLists.txt -x tags wolfssl-3.14.0-pristine/src/ssl.c wolfssl-3.14.0/src/ssl.c
--- wolfssl-3.14.0-pristine/src/ssl.c	2018-03-02 22:39:08.000000000 +0100
+++ wolfssl-3.14.0/src/ssl.c	2018-05-22 16:51:04.794258756 +0200
@@ -14948,7 +14948,7 @@
      *
      * Returns the error value
      */
-    unsigned long wolfSSL_ERR_get_error_line_data(const char** file, int* line,
+    int _INT_wolfSSL_ERR_get_error_line_data(const char** file, int* line,
                                                   const char** data, int *flags)
     {
         WOLFSSL_STUB("wolfSSL_ERR_get_error_line_data");
@@ -14963,6 +14963,24 @@
     }
 
 
+	/* Similar to wolfSSL_ERR_get_error_line but takes in a flags argument for
+	* more flexability.
+	*
+	* file  output pointer to file where error happened
+	* line  output to line number of error
+	* data  output data. Is a string if ERR_TXT_STRING flag is used
+	* flags bit flag to adjust data output
+	*
+	* Returns the error value
+	*/
+	unsigned long wolfSSL_ERR_get_error_line_data(const char** file, int* line,
+		const char** data, int *flags)
+	{
+		int retVal = _INT_wolfSSL_ERR_get_error_line_data(file, line, data, flags);
+		if (retVal < 0) return 0;
+		return retVal;
+	}
+
     WOLFSSL_API pem_password_cb* wolfSSL_CTX_get_default_passwd_cb(
                                                                WOLFSSL_CTX *ctx)
     {
@@ -25556,6 +25574,230 @@
 #endif
     return WOLFSSL_SUCCESS;
 }
+
+
+/* return WOLFSSL_SUCCESS if success, WOLFSSL_FATAL_ERROR if error */
+static int wolfSSL_RSA_PubLoadDer(WOLFSSL_RSA* rsa, const unsigned char* derBuf, int derSz)
+{
+    word32 idx = 0;
+    int    ret;
+
+    WOLFSSL_ENTER("wolfSSL_RSA_LoadDer");
+
+    if (rsa == NULL || rsa->internal == NULL || derBuf == NULL || derSz <= 0) {
+        WOLFSSL_MSG("Bad function arguments");
+        return WOLFSSL_FATAL_ERROR;
+    }
+
+    ret = wc_RsaPublicKeyDecode(derBuf, &idx, (RsaKey*)rsa->internal, derSz);
+    if (ret < 0) {
+        WOLFSSL_MSG("RsaPublicKeyDecode failed");
+        return WOLFSSL_FATAL_ERROR;
+    }
+
+    if (SetRsaExternal(rsa) != WOLFSSL_SUCCESS) {
+        WOLFSSL_MSG("SetRsaExternal failed");
+        return WOLFSSL_FATAL_ERROR;
+    }
+
+    rsa->inSet = 1;
+
+    return WOLFSSL_SUCCESS;
+}
+
+/* returns WOLFSSL_RSA object on success, NULL on failure */
+static RSA * wolfssl_PEM_read_mem_RSA_PUBKEY(void * buffer, int buffer_size)
+{
+    byte *der_buff = NULL;
+    int der_buff_size = 0;
+    int load_result = 0;
+    WOLFSSL_RSA *pub_rsa = NULL;
+
+    if (!buffer || !buffer_size)
+        return NULL;
+
+    if(!(der_buff = (byte *)XMALLOC(buffer_size, NULL, DYNAMIC_TYPE_DER)))
+        return NULL;
+
+    /* Convert public key from PEM format to DER format */
+    der_buff_size = wolfSSL_PubKeyPemToDer((unsigned char *)buffer, buffer_size,
+                                            der_buff, buffer_size);
+
+    if (der_buff_size <= 0)
+    {
+        ForceZero(der_buff, (word32)buffer_size);
+        XFREE(der_buff, NULL, DYNAMIC_TYPE_DER);
+        return NULL;
+    }
+
+    if (!(pub_rsa = wolfSSL_RSA_new()))
+    {
+        ForceZero(der_buff, (word32)buffer_size);
+        XFREE(der_buff, NULL, DYNAMIC_TYPE_DER);
+        return NULL;
+    }
+
+    /* Create RSA structure from public key in DER format */
+    load_result = wolfSSL_RSA_PubLoadDer(pub_rsa, der_buff, der_buff_size);
+    if (load_result == WOLFSSL_FATAL_ERROR)
+    {
+        ForceZero(der_buff, (word32)buffer_size);
+        XFREE(der_buff, NULL, DYNAMIC_TYPE_DER);
+        wolfSSL_RSA_free(pub_rsa);
+        return NULL;
+    }
+
+    return pub_rsa;
+}
+
+WOLFSSL_RSA *wolfssl_PEM_read_bio_RSA_PUBKEY(WOLFSSL_BIO *bp, WOLFSSL_RSA **x,
+                                             pem_password_cb *cb, void *u)
+{
+    byte *pem_buf;
+    int pem_buf_size;
+
+    if (bp == NULL) {
+        WOLFSSL_MSG("NULL argument passed in");
+        return NULL;
+    }
+    if (bp->type != WOLFSSL_BIO_MEMORY) {
+        WOLFSSL_MSG("currently only works with memory BIOs");
+        return NULL;
+    }
+    pem_buf_size = wolfSSL_BIO_get_mem_data(bp, &pem_buf);
+    /* Create RSA structure from DER buffer */
+    return wolfssl_PEM_read_mem_RSA_PUBKEY(pem_buf, pem_buf_size);
+}
+
+WOLFSSL_RSA *wolfssl_PEM_read_RSA_PUBKEY(FILE * fp,
+                                 WOLFSSL_RSA ** x,
+                                 pem_password_cb * cb,
+                                 void * u)
+{
+    int file_size = 0;
+    int pem_buff_size = 0;
+    byte *pem_buff = NULL;
+    XFILE file = (XFILE) fp;
+    WOLFSSL_RSA *pub_rsa = NULL;
+
+    (void)x;
+    (void)cb;
+    (void)u;
+    /* Get file size */
+    if (!file)
+        return NULL;
+    XFSEEK(file, 0, XSEEK_END);
+    file_size = XFTELL(file);
+    XREWIND(file);
+
+    if (!file_size ||
+        !(pem_buff = (byte *)XMALLOC(file_size, NULL, DYNAMIC_TYPE_PEM)))
+        return NULL;
+
+    /* Read file content */
+    pem_buff_size = (int) XFREAD(pem_buff, 1, file_size, file);
+    if (pem_buff_size != file_size)
+    {
+        ForceZero(pem_buff, (word32)file_size);
+        XFREE(pem_buff, NULL, DYNAMIC_TYPE_PEM);
+        return NULL;
+    }
+
+    /* Create RSA structure from DER buffer */
+    if (!(pub_rsa = wolfssl_PEM_read_mem_RSA_PUBKEY(pem_buff, pem_buff_size)))
+    {
+        ForceZero(pem_buff, (word32)file_size);
+        XFREE(pem_buff, NULL, DYNAMIC_TYPE_PEM);
+        return NULL;
+    }
+
+    return pub_rsa;
+}
+
+WOLFSSL_RSA *wolfssl_PEM_read_RSAPrivateKey(FILE *fp,
+                                    WOLFSSL_RSA ** x,
+                                    pem_password_cb * cb,
+                                    void * u)
+{
+    int file_size = 0;
+    int pem_buff_size = 0;
+    int der_buff_size = 0;
+    int load_result = 0;
+    byte *pem_buff = NULL;
+    byte *der_buff = NULL;
+    WOLFSSL_RSA *priv_rsa = NULL;
+    XFILE file = (XFILE) fp;
+
+    (void)x;
+    (void)cb;
+    (void)u;
+    /* Get file size */
+    if (!file)
+        return NULL;
+    XFSEEK(file, 0, XSEEK_END);
+    file_size = XFTELL(file);
+    XREWIND(file);
+
+    if (!file_size ||
+        !(pem_buff = (byte *)XMALLOC(file_size, NULL, DYNAMIC_TYPE_PEM)))
+    {
+        return NULL;
+    }
+
+    /* Read file content */
+    pem_buff_size = (int) XFREAD(pem_buff, 1, file_size, file);
+    if (pem_buff_size != file_size)
+    {
+        ForceZero(pem_buff, (word32)file_size);
+        XFREE(pem_buff, NULL, DYNAMIC_TYPE_PEM);
+        return NULL;
+    }
+
+    if (!(der_buff = (byte *)XMALLOC(pem_buff_size, NULL, DYNAMIC_TYPE_DER)))
+    {
+
+        ForceZero(pem_buff, (word32)file_size);
+        XFREE(pem_buff, NULL, DYNAMIC_TYPE_PEM);
+        return NULL;
+    }
+
+    /* Convert private key from PEM format to DER format */
+    der_buff_size = wolfSSL_KeyPemToDer(pem_buff, pem_buff_size,
+                                        der_buff, pem_buff_size,
+                                        NULL);
+
+    if (der_buff_size <= 0)
+    {
+        ForceZero(pem_buff, (word32)file_size);
+        XFREE(pem_buff, NULL, DYNAMIC_TYPE_PEM);
+        ForceZero(der_buff, (word32)pem_buff_size);
+        XFREE(der_buff, NULL, DYNAMIC_TYPE_DER);
+        return NULL;
+    }
+
+    if (!(priv_rsa = wolfSSL_RSA_new()))
+    {
+        ForceZero(pem_buff, (word32)file_size);
+        XFREE(pem_buff, NULL, DYNAMIC_TYPE_PEM);
+        ForceZero(der_buff, (word32)pem_buff_size);
+        XFREE(der_buff, NULL, DYNAMIC_TYPE_DER);
+        return NULL;
+    }
+
+    /* Create RSA structure from private key in DER format */
+    load_result = wolfSSL_RSA_LoadDer(priv_rsa, der_buff, der_buff_size);
+    if (load_result == WOLFSSL_FATAL_ERROR)
+    {
+        ForceZero(pem_buff, (word32)file_size);
+        XFREE(pem_buff, NULL, DYNAMIC_TYPE_PEM);
+        ForceZero(der_buff, (word32)pem_buff_size);
+        XFREE(der_buff, NULL, DYNAMIC_TYPE_DER);
+        wolfSSL_RSA_free(priv_rsa);
+        return NULL;
+    }
+
+    return priv_rsa;
+}
 #endif /* NO_RSA */
 
 #ifndef NO_WOLFSSL_STUB
@@ -31991,7 +32233,7 @@
 #endif
 #endif
 
-#if defined(WOLFSSL_NGINX) || defined(WOLFSSL_HAPROXY)
+#if defined(WOLFSSL_NGINX) || defined(WOLFSSL_HAPROXY) || defined(OPENSSL_EXTRA)
 int wolfSSL_X509_get_ex_new_index(int idx, void *arg, void *a, void *b, void *c)
 {
     static int x509_idx = 0;
@@ -32191,6 +32433,22 @@
     return WOLFSSL_SUCCESS;
 }
 
+
+
+int wolfSSL_X509_check_ip_asc(WOLFSSL_X509 *x, const char *address, unsigned int flags)
+{
+  (void) x;
+  (void) address;
+  (void) flags;
+
+#ifdef WOLFSSL_MYSQL_COMPATIBLE
+    return 1;
+#else
+    return WOLFSSL_NOT_IMPLEMENTED;
+#endif
+}
+
+
 int wolfSSL_i2a_ASN1_INTEGER(BIO *bp, const WOLFSSL_ASN1_INTEGER *a)
 {
     static char num[16] = { '0', '1', '2', '3', '4', '5', '6', '7',
diff -ur -N -d -x CMakeLists.txt -x tags wolfssl-3.14.0-pristine/wolfcrypt/src/asn.c wolfssl-3.14.0/wolfcrypt/src/asn.c
--- wolfssl-3.14.0-pristine/wolfcrypt/src/asn.c	2018-03-02 22:39:08.000000000 +0100
+++ wolfssl-3.14.0/wolfcrypt/src/asn.c	2018-05-22 16:24:14.725988085 +0200
@@ -4294,9 +4294,11 @@
                 #endif /* OPENSSL_EXTRA */
             }
             if (copy && !tooBig) {
-                XMEMCPY(&full[idx], copy, XSTRLEN(copy));
-                idx += (word32)XSTRLEN(copy);
-                XMEMCPY(&full[idx], &cert->source[cert->srcIdx], strLen);
+				XMEMCPY(&full[idx], copy, XSTRLEN(copy));
+				idx += (word32)XSTRLEN(copy);
+				XMEMCPY(&full[idx], "=", XSTRLEN("="));
+				idx += (word32)XSTRLEN("=");
+				XMEMCPY(&full[idx], &cert->source[cert->srcIdx], strLen);
                 idx += strLen;
             }
 
diff -ur -N -d -x CMakeLists.txt -x tags wolfssl-3.14.0-pristine/wolfssl/openssl/pem.h wolfssl-3.14.0/wolfssl/openssl/pem.h
--- wolfssl-3.14.0-pristine/wolfssl/openssl/pem.h	2018-03-02 22:39:08.000000000 +0100
+++ wolfssl-3.14.0/wolfssl/openssl/pem.h	2018-05-22 16:24:14.725988085 +0200
@@ -59,10 +59,19 @@
 WOLFSSL_RSA *wolfSSL_PEM_read_RSAPublicKey(FILE *fp, WOLFSSL_RSA **x,
                                            pem_password_cb *cb, void *u);
 WOLFSSL_API
+WOLFSSL_RSA *wolfssl_PEM_read_RSAPrivateKey(FILE *, WOLFSSL_RSA **, pem_password_cb *, void *);
+
+WOLFSSL_API
+WOLFSSL_RSA *wolfssl_PEM_read_bio_RSA_PUBKEY(WOLFSSL_BIO *bp, WOLFSSL_RSA **x,
+                                             pem_password_cb *cb, void *u);
+WOLFSSL_API
 int wolfSSL_PEM_write_RSAPublicKey(FILE *fp, WOLFSSL_RSA *x);
 
 WOLFSSL_API
 int wolfSSL_PEM_write_RSA_PUBKEY(FILE *fp, WOLFSSL_RSA *x);
+
+WOLFSSL_API
+WOLFSSL_RSA * wolfssl_PEM_read_RSA_PUBKEY(FILE *, WOLFSSL_RSA **, pem_password_cb *, void *);
 #endif /* NO_FILESYSTEM */
 
 /* DSA */
@@ -148,6 +157,7 @@
 #define PEM_write_RSA_PUBKEY        wolfSSL_PEM_write_RSA_PUBKEY
 #define PEM_write_RSAPublicKey      wolfSSL_PEM_write_RSAPublicKey
 #define PEM_read_RSAPublicKey       wolfSSL_PEM_read_RSAPublicKey
+#define PEM_read_bio_RSA_PUBKEY     wolfssl_PEM_read_bio_RSA_PUBKEY
 /* DSA */
 #define PEM_write_bio_DSAPrivateKey wolfSSL_PEM_write_bio_DSAPrivateKey
 #define PEM_write_DSAPrivateKey     wolfSSL_PEM_write_DSAPrivateKey
@@ -161,6 +171,9 @@
 #define PEM_read_PUBKEY         wolfSSL_PEM_read_PUBKEY
 #define EVP_PKEY_type           wolfSSL_EVP_PKEY_type
 
+#define PEM_read_RSAPrivateKey  wolfssl_PEM_read_RSAPrivateKey
+#define PEM_read_RSA_PUBKEY     wolfssl_PEM_read_RSA_PUBKEY
+
 #ifdef __cplusplus
     }  /* extern "C" */ 
 #endif
diff -ur -N -d -x CMakeLists.txt -x tags wolfssl-3.14.0-pristine/wolfssl/openssl/rsa.h wolfssl-3.14.0/wolfssl/openssl/rsa.h
--- wolfssl-3.14.0-pristine/wolfssl/openssl/rsa.h	2018-03-02 22:38:24.000000000 +0100
+++ wolfssl-3.14.0/wolfssl/openssl/rsa.h	2018-05-22 16:24:14.726988101 +0200
@@ -43,6 +43,7 @@
     #define RSA_PKCS1_PADDING      0
     #define RSA_PKCS1_OAEP_PADDING 1
 #else
+    #include <wolfssl/wolfcrypt/rsa.h> // for the following definitions
     #define RSA_PKCS1_PADDING      WC_RSA_PKCSV15_PAD
     #define RSA_PKCS1_OAEP_PADDING WC_RSA_OAEP_PAD
 #endif /* HAVE_FIPS */
diff -ur -N -d -x CMakeLists.txt -x tags wolfssl-3.14.0-pristine/wolfssl/openssl/ssl.h wolfssl-3.14.0/wolfssl/openssl/ssl.h
--- wolfssl-3.14.0-pristine/wolfssl/openssl/ssl.h	2018-03-02 22:39:08.000000000 +0100
+++ wolfssl-3.14.0/wolfssl/openssl/ssl.h	2018-05-22 16:43:34.236077238 +0200
@@ -874,6 +874,7 @@
 #define SSL_in_init                       wolfSSL_SSL_in_init
 #define SSL_get0_session                  wolfSSL_SSL_get0_session
 #define X509_check_host                   wolfSSL_X509_check_host
+#define X509_check_ip_asc                 wolfSSL_X509_check_ip_asc
 #define i2a_ASN1_INTEGER                  wolfSSL_i2a_ASN1_INTEGER
 #define ERR_peek_error_line_data          wolfSSL_ERR_peek_error_line_data
 #define ERR_load_BIO_strings              wolfSSL_ERR_load_BIO_strings
diff -ur -N -d -x CMakeLists.txt -x tags wolfssl-3.14.0-pristine/wolfssl/ssl.h wolfssl-3.14.0/wolfssl/ssl.h
--- wolfssl-3.14.0-pristine/wolfssl/ssl.h	2018-03-02 22:39:08.000000000 +0100
+++ wolfssl-3.14.0/wolfssl/ssl.h	2018-05-22 16:50:55.805110309 +0200
@@ -2727,7 +2727,7 @@
 WOLFSSL_API void wolfSSL_OPENSSL_config(char *config_name);
 #endif
 
-#if defined(WOLFSSL_NGINX) || defined(WOLFSSL_HAPROXY)
+#if defined(WOLFSSL_NGINX) || defined(WOLFSSL_HAPROXY) || defined(OPENSSL_EXTRA)
 /* Not an OpenSSL API. */
 WOLFSSL_LOCAL int wolfSSL_get_ocsp_response(WOLFSSL* ssl, byte** response);
 /* Not an OpenSSL API. */
@@ -2760,6 +2760,8 @@
 #endif
 WOLFSSL_API int wolfSSL_X509_check_host(WOLFSSL_X509 *x, const char *chk,
     size_t chklen, unsigned int flags, char **peername);
+WOLFSSL_API int wolfSSL_X509_check_ip_asc(WOLFSSL_X509 *x, const char *address,
+     unsigned int flags);
 
 WOLFSSL_API int wolfSSL_i2a_ASN1_INTEGER(WOLFSSL_BIO *bp,
     const WOLFSSL_ASN1_INTEGER *a);
